### PR TITLE
cSpell checking json and code

### PR DESCRIPTION
**Pull request by isaec**
This pr adds [cSpell](https://cspell.org/) configuration which should be a reasonable starting place. cSpell is a spellchecker intended for use on code. It has a [cli](https://www.npmjs.com/package/cspell), [vscode extension](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker), and can be configured with [dictionaries](https://cspell.org/docs/dictionaries/).

This pr adds a dictionary for all the (flagged as typo) words in the ``*.json`` files inside ``/data``, and a dictionary for words in code. You can look in the files changed to get a feel for it - words like "Loadbrew" which are not typos in the context of 5etools, but are typos normally.

My intent was that javascript-words would be code dictionary, and dnd-words would be data dictionary - making it hypothetically possible for the dnd-words dictionary to be manually checked, although this seems like far too much work to me.

Additionally, it adds two npm commands to maintain these lists, ``npm run spellcheck:data``, which spellchecks all the json inside the data directory, and ``npm run spellcheck:trust-data``, which in my mind is used when new data is added with trusted spelling, or to generate a git diff with the new, possibly misspelled words.

I am not married to the naming of these commands, but I had a hard time coming up with something.

The way cSpell works is by highlighting words that aren't on any of the cSpell dictionary lists. It rarely has false positives in my experience, and unlike most spell checkers it's intended for use on code, so it doesn't trip for syntax.

If you have the cSpell extension installed, you can now see typos highlighted in code. I generated a javascript dictionary of all the words cSpell thought were typos in the non-minified code I saw in the repo - I could have missed a spot, I don't know this code.

image
image

Using the quick actions is the way cSpell is intended to be used for code - so if new code is written and cSpell highlights something as a typo, it can be marked as intentional in the javascript dictionary. 

More interestingly, I think this can also help with work on the json files that make the data for 5etools. I generated a dnd dictionary of all the words cSpell flagged in the data directory - there may be typos, but now there is some level of intellisense when editing these files.
image
image
These examples are pretty contrived, but I still think it has some value. Running from the cli, it looks like this:
scr_20:16:34
On my laptop, which is fairly fast, it takes over a minute to spellcheck all the data fwiw.

Because it is possible for such a large amount of data to be added - say, a new book - I also added the ``npm run spellcheck:trust-data`` command, which will scan all the data files for perceived typos, and then add them to the dictionary as words. 

Finally, this pr checks in a settings.json and extensions.json that let a vscode user know that these extensions are helpful. This is shown as a little popup when a repository is first opened (I think).

The settings.json also enforces what appears to me to be the styling being used in the project - I had prettier installed, and without this change I would make massive diffs if I saved a file from reformatting. This settings.json makes sure vscode users format with eslint, and don't autoformat other files - I assume a formatter is used for the json files, but I couldn't figure out which. While I think it's good to have the styling settings committed to vc, I wouldn't object to removing them from this PR.

Thanks for your time and all you do for this project! Hopefully this isn't entirely unwanted.

### CODE REVIEW

**Comment by isaec** [comment]
It is [possible](https://www.npmjs.com/package/cspell#:~:text=Git%3A%20Check%20Only%20Changed%20Files) to use git to only spellcheck *new* data files, but I didn't see the need - if that's desired I can add it.

**Comment by revilowaldow** [comment]
I'm not for this change. It will cause direct conflicts with anyone who already has a .VScode directory that is currently setup for them. 
IDE files are personal to the user and an alternative method should be used to store spell checking information.
More specifically you also seem to have set a new default formatting extension rather than using the incumbent npm script to clean jsons. 
For what it's worth, I also use cSpell, but I don't think it's inclusion is worth the user impact from adding an IDE settings file to the git history. 
Also, you haven't submitted any changes to the data, I assume that means it hasn't found any issues?

**Comment by TheGiddyLimit** [comment]
Agree with the above.

Additionally, not everyone uses VSCode (I use JetBrains offerings, for example, so this is worthless to _me_ as-is)--if the non-VSCode-specific parts could be split out then we could probably offer a simple setup script, or even just a README note, about setting up spellcheck on various IDEs. I see no reason the dictionary files need to be kept in a VSCode-specific directory?

 I'm not expecting this PR to expand beyond VSCode, just that what's here should be made IDE-agnostic.

Adding an un-reviewed list of 20k+ words is also a _bit_ sketchy 🥴

**Comment by isaec** [comment]
Thanks for the quick replies. I should have kept the settings and extensions things independent. 

The change to the formatting in ``settings.json`` is for the editor - I can't speak for everyone, but my vscode is set up to run prettier on save for any file, which makes massive and undesired reformatting. A note in the repo that eslint is how formatting for code should be set up in the readme is better then nothing, but worse then just shipping said eslint vscode config imho. Revilowaldow, if you have a vscode config that works for you, I would be in favor of checking that in instead - anything that lets me get the same formatting experience when editing code that the people who *review* code want, without guessing at the intent.

The changes to ``settings.json`` are about what vscode does when saving a ``.js`` file, and don't do any kind of formatting for jsons - only for code. It is an edit time code formatting, not an alternative to the npm script for formatting json.

I probably should have been more clear about how I generated these dictionaries. I put the command I used in the commit message: ``npx cspell --words-only --unique ./data/**/*.json | sort --ignore-case >> .vscode/dnd-words.txt``. What this command does is use the cspell cli to find all the words it thinks are typos, and then runs em through sort and writes em to the dictionary test file. This isn't a change that would catch current spelling errors unless someone wants to review the 20k words it took issue with - it would work on the assumption that the number of typos are low in the current data, and future data could be checked against this.

If a typo is found in current data somewhere, you could check if it was in the cSpell dictionary. If it was, you would find it elsewhere in data (not any better then a ripgrep) but also highlight that typo in the future.

If you want to "verify" the origin of the words, you can delete everything in the ``dnd-words.txt`` and run that command - assuming there are no race conditions in the cspell code, it **should** make an identical file.

As for the location of files - there is no reason dictionaries need to live in .vscode (my bad), and the cspell json can also be placed in project root. I can move the cspell config to root, and put the dictionaries in a folder - how do i configure nginx to not serve these files?

As for the utility for non-vscode editors, the cli still works, and could be put into a github commit check or similar, but that's way beyond what I was originally thinking.

In jetbrains, if you wanted to check new files, you could run ``npm run spellcheck:data`` (bad name) and it would list what it thinks are typos in the folders. In the scenario where you just added a new book, it would spit out a large list of proper nouns that are new to the book. My thinking was it would be within reason to have a few people look at 1k words and cross reference them with their location in the book to see if they were new words or typos.

I could add an npm command to check code, but I imagine that is undesired.

I can close the pr, or I can drop the editor code formatting changes, and move cSpell stuff to it's own directory + root.

I think you should consider the usability of not including much beyond hints as to how code formatting is expected, but I would be happy to add a readme change in a separate pr about how formatting is done for the project.

Hopefully this isn't too rambly, yall gave me a lot of different things to reply to :grin:

**Comment by TheGiddyLimit** [comment]
Fair enough re: `.vscode` formatting/etc., I'll admit I dismissed it out of hand without digging into the specifics [Stack Overflow](https://stackoverflow.com/a/57749909) seems to think a minimal set of JSON files is fine, so what you have there looks good. All you VSCode users can fight out the specifics 😏 

> move cSpell stuff to it's own directory + root

that would be ideal. No strong opinions on what one would name that, here--`spellcheck/*`?

The rambling is all helpful, no worries!

**Comment by isaec** [comment]
``/spellcheck`` or ``/cspell_dictionary`` or the hyphen equivalent all seem fine by me.

The utility of this kind of spell checking for the data directory is something I can't assess, if checking against previous words for _new_ strange spellings doesn't actually help I see no need to merge this. My intuition was that it would catch a subset of spelling errors in the future.

@revilowaldow, what's in the project specific settings on your machine?

**Comment by revilowaldow** [comment]
> what's in the project specific settings on your machine?
```json
{
	"json.schemas": [
		{
			"fileMatch": [
				"*.json"
			],
			"url": "../5etools-mirror-1.github.io/test/schema/homebrew.json"
		}
	],
	"json.schemaDownload.enable": true,
	"explorer.experimental.fileNesting.enabled": true,
	"explorer.experimental.fileNesting.patterns": {
		"*.ts": "${capture}.js",
		"*.js": "${capture}.js.map, ${capture}.min.js, ${capture}.d.ts",
		"*.jsx": "${capture}.js",
		"*.tsx": "${capture}.ts",
		"tsconfig.json": "tsconfig.*.json",
		"package.json": "package-lock.json, yarn.lock",
		"index.html": "*.html",
		"README.md": "*.md",
		"*.json": "fluff-${capture}.json, foundry-${capture}.json, roll20-${capture}.json",
		"apple-touch-icon-360x360.png": "*.png",
		"filter.js": "filter*.js",
		"converter.js": "converter*.js",
		"dmscreen.js": "dmscreen*.js",
		"render.js": "render*.js"
	},
	"files.maxMemoryForLargeFilesMB": 8192,
	"editor.wordWrap": "on",
	"editor.bracketPairColorization.enabled": true,
	"editor.bracketPairColorization.independentColorPoolPerBracketType": false,
	"editor.guides.bracketPairsHorizontal": true,
	"editor.foldingMaximumRegions": 10000,
	"editor.cursorSmoothCaretAnimation": false,
	"editor.suggest.showStatusBar": true,
	"cSpell.language": "en-GB",
	"cSpell.suggestionsTimeout": 4000,
	"gitlens.advanced.repositorySearchDepth": 1,
	"json.trace.server": "verbose"
}
```
Some of these, but definitely not all, could do with being moved up a level.
My point is that `.vscode/settings.json` is a personal user preference file, and not something we should be including in the repo.

Additionally, if this [extension](https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.triggertaskonsave) is installed then putting the following in your *workspace* settings (a level up) will happily clean the jsons for you using the supplied npm script.
```json
"triggerTaskOnSave.tasks": {
	"clean-jsons": [
		"*.json"
	]
}
```

**Comment by isaec** [comment]
The changes in this pr in regards to json are about spell checking, not about formatting - nothing in this pr changes json cleaning.

The filenesting patterns in your ``settings.json`` are **brilliant**, and something I would strongly advocate for checking into version control - I didn't know vscode even had this functionality, but it definitely helps one move around this repo. It makes a massive difference to the experience of opening this repo - things are logically grouped! 

While I was aware of the json schema functionality, I wasn't aware one existed for this project. If you check settings.json into vc with the schema, every user of vscode can get autocompletion in their changes to json - a big productivity and usability improvement in my opinion.

[``.vscode/settings.json`` _is_ the workspace settings file.](https://code.visualstudio.com/docs/getstarted/settings#_workspace-settingsjson-location) It is intended for settings that should be shared between people who participate on the project. User specific settings should exist in [user settings](https://code.visualstudio.com/docs/getstarted/settings#_settingsjson) - that is where personal settings go. This may not be how this project is used, and I can accept that - but ``.vscode/settings.json`` is, according to vscode docs, intended to be shared.

If this is a reference to ``.code-workspace`` files and I misunderstood, those are intended for monorepos / multiroot workspaces.
https://stackoverflow.com/questions/44629890/what-is-a-workspace-in-visual-studio-code

I would agree that stuff like bracket colorization is personal - I don't particularly care for it. The vscode settings model assumes that personal preferences like that are per language, so you can language scope your bracket settings in user settings. I feel the same about editor wrapping, memory limits ect - they aren't actually specific to this project.

I can't tell if my tone is coming across correctly, I don't mean any disrespect and I appreciate the work yall do.

**Comment by revilowaldow** [comment]
## Damaging repercussions
It needs to not break everyone's vscode setup, ***adding a new vscode settings file will irretrievably delete anyone else's settings as theirs will not have been subject to version control and an inadvertent merge/pull will overwrite them.***

## VSCode Settings sit at 4 levels:
image
User, Machine, Workspace and folder. The file you are editing is the folder settings, and those effectively become the workspace settings if you only have this folder open, many people run it in multiroot with the homebrew repo for example.

## Cleaning/Style
This is your change that deals with cleaning the files. 
image
If you don't want Prettier breaking the formatting every time, then that's on you to turn off (or disable for this folder) prettier on your machine, the repo already has a fully featured linting script included, and the eslint definitions are in `node\.eslint.js`

Also, this is the only change you're making to `.vscode/settings.json`, if we eliminate this, then the whole scenario described above is avoided.

The script is only one extra click away from formatting the document.
image
image

## My Settings

- Yes the file arranging is pretty cool, but also, I don't want to commit anything with `experimental` in the name.

- I see you've changed the directory on the schema `fileMatch`, I'll add a warning that that is particularly fucky and multiple people in the server have had issues with it. It does not always do what you expect it to do repeatably. 
For example see https://github.com/microsoft/vscode/issues/147865 
- The schema is great, but also, it's barely ever used on the site data files except by Giddy, who actually uses the npm scripts not the live validation.
The vast majority of people use it in the homebrew repo, which has it's own more permissive schema, and most people there simply add this to the start of their single file for the same effect.
```json
"$schema": "https://raw.githubusercontent.com/TheGiddyLimit/homebrew/master/_schema-fast/homebrew.json",  
```
## The list of 21,281 words needs to be checked over, or it'll never be used.
Giddy has already said this doesn't work for his setup so he's not likely to use it, if you the pull requester, are not prepared to go through the list, then I don't imagine anyone else would be.
And if ultimately, only you will use this, then it's better to keep it in your own fork. 
Commits with fixes to the data are always welcome.

Additionally, cSpell has it's own problems. 
image

## A note
Similarly to what you've said above, I don't want to come across with the wrong tone, thank you for the pull request, it's appreciated. I have specific concerns I've mentioned above about changing the settings file.
If this was a new project then I'd be all for starting strong, but it's now a significantly mature project with nearly 200 (visible) forks of just this mirror. Changing the file at this late stage has the potential to be damaging. We should focus in on the spel checking aspect of this and not the vscode settings.

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
+{
+  // code formatting section
```
Comment (revilowaldow) [inline .vscode/settings.json:2]: Comments in json are only supported in the vscode specific `.jsonc` file type. Please fix.

Comment (isaec) [reply .vscode/settings.json:2]: Would you like me to use the ``.jsonc`` format, or remove the comments? VSCode will play just fine with the ``.json`` extension, but I guess other editors would not.

**Comment by isaec** [comment]
## in regards to settings.json
I think I just understood what you are saying about ``settings.json`` files overwriting custom stuff. Sorry I didn't get that - because your file is not checked in, if you pulled down this branch you would nuke what you have, irreparably. Thank you for preventing that lmao, I genuinely didn't understand.

Do you think ``.code-workspace`` is less likely to be already in use? It is not present in gitignore, so a user couldn't already have custom things in there - but I definitely choose a bad first pr, I don't fully understand how this repo is used.

The schema bug is interesting - my gut is that its an overly specific glob, but it may actually be a vscode bug. 

I will remove the ``.vscode`` changes from the pr, I have been convinced lmao.

## in regards to the spellchecking
I disagree that this spellchecking dictionary needs to be manually checked over to be useful. In the example you give, this branch does not feature the erroneously flagged typo. Of course it doesn't, the dictionary of 20k words contains everything cspell thought was a typo!
image
But, it can still catch _actual_ typos:
image

Obviously, one would need to make sure every word in that list is a real dnd word if you wanted to catch every typo - but a lot of them are names, it would take eons.

I still think it buys a fair amount for a tiny amount of effort to check for _new_ typos on new files. If you, Giddy disagree I will close the pr.

The cli  ``npm run spellcheck:data`` version of this would work for giddy's setup. I can add a command to check faster (only validate the modified files) to make that more ergonomic.

**Comment by isaec** [comment]
This definitely needs a squash, but it should only be spellchecking changes now.

``npm run`` ``spellcheck:check-data``, ``spellcheck:check-data-quick`` and ``spellcheck:trust-data`` are what I have changed the command names to. The quick variant uses uses regex and git to determine what json data files have been changed, and only spellcheck those.

It also looks like there are jetbrains extensions for this type of spellchecking dictionary, but that is a can of worms I will not be opening.

I do question the utility / need of this change. Should I close the pr?

**Comment by TheGiddyLimit** [comment]
> I do question the utility / need of this change

nothing here requires real maintenance, so I'm happy to have it just for the sake of having it

**Review by TheGiddyLimit — APPROVED** [review]

**Comment by revilowaldow** [comment]
Yeah with the switch to just the spell check, and the elimination of the vscode changes I'm satisfied as well, (not that that really means anything) 😂 thanks for the contribution!

**Comment by TheGiddyLimit** [comment]
Merged to main repo for next release.

One thing to note, that I'll probably end up changing, is that the package script naturally don't work on Windows due to all the POSIX(?)-specific syntax/tools. These should really be platform-agnostic, because I don't want change which shell I'm using to run a spellcheck command!

**Comment by isaec** [comment]
I genuinely forgot that people develop without a posix shell. I don't know how to do that sort of operation without a shell, or more code - maybe inline javascript?

Regardless, ty and good luck.

**Pull request closed without merging**